agent/lib/kontena/observer.rb

```diff
@@ -1,117 +1,335 @@
+    # Set value for observable
+    #
+    # @raise [RuntimeError] unknown observable
+    # @return value
+    def set(observable, value)
+      raise "unknown observable: #{observable.class.name}" unless @values.has_key? observable
+
+      @values[observable] = value
+    end
+
+    # Update observed value from message
+    #
+    # @param message [Kontena::Observable::Message]
+    def update(message)
+      debug { "update #{message.observable} -> #{message.value}" }
+
+      set(message.observable, message.value)
+    end
+
+    # Return next observable messages sent to this actor from Observables using #<<
+    # Suspends the calling celluloid task in between message yields.
+    # Must be called atomically, suspending in between calls to receive() risks missing intervening messages!
+    #
+    # @raise [Timeout::Error]
+    def receive
+      timeout = @deadline ? @deadline - Time.now : nil
+
+      debug { "receive timeout=#{timeout}..." }
+
+      begin
+        # Celluloid.receive => Celluloid::Actor#receive => Celluloid::Internals::Receiver#receive returns nil on timeout
+        message = Celluloid.receive(timeout) { |msg| Kontena::Observable::Message === msg && msg.observer == self }
+      rescue Celluloid::TaskTimeout
+        # Celluloid.receive => Celluloid::Mailbox.receive raises TaskTimeout insstead
+        message = nil
+      end
+
+      if message
+        return message
+      else
+        raise Timeout::Error, "observe timeout #{'%.2fs' % timeout}: #{self.describe_observables}"
       end
+    end
+
+    # Any observable has an error?
+    #
+    # @return [Boolean] true => some observed value is an Exception
+    def error?
+      @values.any? { |observable, value| Exception === value }
+    end
+
+    # Every observable has a value?
+    #
+    # @return [Boolean] false => some observed values are still nil
+    def ready?
+      !@values.any? { |observable, value| value.nil? }
+    end
+
+    # Map each observed observable to its value.
+    #
+    # Should only be used once ready?
+    #
+    # @return [Array] observed values
+    def values
+      @observables.map{|observable| @values[observable] }
+    end
 
-      # immediate async update if all observables were ready
-      async.observed(observe)
+    # Return Error for first crashed observable.
+    #
+    # Should only be used if error?
+    #
+    # @return [Exception, nil]
+    def error
+      @values.each_pair{|observable, value|
+        return Error.new(observable, value) if Exception === value
+      }
+      return nil
+    end
 
-      observe
+    # Yield each set of ready? observed values while alive, or raise on error?
+    #
+    # The yield is exclusive, because suspending the observing task would mean that
+    # any observable messages would get discarded.
+    #
+    # @param timeout [Float] timeout between each yield
+    def each(timeout: nil)
+      @deadline = Time.now + timeout if timeout
+
+      while alive?
```

Comment:
> Meaning that if I `kill` my observer, it might still get an update as it could have been receiving?

This `while alive?` loop is kinda cosmetic... there is no known case where that would ever be false, because the observer gets killed by the `ensure` block in the `Kontena::Observer.observe` method calling this function.

This should be changed to `while true`?